### ISSUE TITLE
使用PyTorch实现DGN时遇到的一处细节问题

### ISSUE THREAD

**Issue by wangwwno1** [PKU-AI-Edge/DGN#5]
作者你好！我最近在尝试用PyTorch将DGN的Relation Kernel迁移到DDPG算法上
想请教一处细节问题，就是MultiHeadsAttModel的代码中，有一个叫ve的输入
ve看起来是个一维的One-Hot向量，且第一个元素为1，请问这个输入的意义和作用是什么？

附上原始代码：
```
def MultiHeadsAttModel(l=2, d=128, dv=16, dout=128, nv = 8 ):

    v1 = Input(shape = (l, d))
    q1 = Input(shape = (l, d))
    k1 = Input(shape = (l, d))
    ve = Input(shape = (1, l))

    v2 = Dense(dv*nv, activation = "relu",kernel_initializer='random_normal')(v1)
    q2 = Dense(dv*nv, activation = "relu",kernel_initializer='random_normal')(q1)
    k2 = Dense(dv*nv, activation = "relu",kernel_initializer='random_normal')(k1)

    v = Reshape((l, nv, dv))(v2)
    q = Reshape((l, nv, dv))(q2)
    k = Reshape((l, nv, dv))(k2)
    v = Lambda(lambda x: K.permute_dimensions(x, (0,2,1,3)))(v)
    k = Lambda(lambda x: K.permute_dimensions(x, (0,2,1,3)))(k)
    q = Lambda(lambda x: K.permute_dimensions(x, (0,2,1,3)))(q)

    att = Lambda(lambda x: K.batch_dot(x[0],x[1] ,axes=[3,3]) / np.sqrt(dv))([q,k])# l, nv, nv
    # 如果要限制无关联Agent的交流，可以在这里做mask处理
    att = Lambda(lambda x: K.softmax(x))(att)
    out = Lambda(lambda x: K.batch_dot(x[0], x[1],axes=[3,2]))([att, v])
    out = Lambda(lambda x: K.permute_dimensions(x, (0,2,1,3)))(out)

    out = Reshape((l, dv*nv))(out)

    T = Lambda(lambda x: K.batch_dot(x[0],x[1]))([ve,out])

    out = Dense(dout, activation = "relu",kernel_initializer='random_normal')(T)
    model = Model(inputs=[q1,k1,v1,ve], outputs=out)   # ←ve在这里的作用是什么？
    return model
```

**Comment by KK666-AI** [comment]
@wangwwno1 can you reproduce their experiment results as reported in their paper?

**Comment by lylyjy** [comment]
Do you have this code of pytorch version?Can you share it?Thanks.

**Comment by wangwwno1** [comment]
> Do you have this code of pytorch version?Can you share it?Thanks.

Sorry but I didn't keep the code as **I can't reproduce the result of this paper**.
The DGN model is a Graph Attention Network in esstential.  So if you want to implement the structure in PyTorch, [this repo](https://github.com/Diego999/pyGAT) might helps. PyTorch also provide APIs for masked attention mechanism, read [this doc](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html) for details.
For the background knowledge, search for "Attention is all your need" and "Graph Attention Network".
